### PR TITLE
Keep taskRunner's cancelAll safe.

### PR DESCRIPTION
**Pull request by JarvisGG**

### CODE REVIEW

**Review by yschimke — APPROVED** [review]
Looks like an appropriate change.  We generally only call from test runners since it is internal API (by package convention, not by visibility)

Since this is not public API, can you expand on whether it is related to existing behaviour in tests or you need this for some particular reason?

**Comment by JarvisGG** [comment]
> Looks like an appropriate change. We generally only call from test runners since it is internal API (by package convention, not by visibility)
> 
> Since this is not public API, can you expand on whether it is related to existing behaviour in tests or you need this for some particular reason?

Just found out this method was a risk operation.😄

File: okhttp/src/main/kotlin/okhttp3/internal/concurrent/TaskRunner.kt [hunk 1]
```diff
@@ -243,14 +243,16 @@ class TaskRunner(
...
   fun cancelAll() {
-    for (i in busyQueues.size - 1 downTo 0) {
-      busyQueues[i].cancelAllAndDecide()
-    }
-    for (i in readyQueues.size - 1 downTo 0) {
-      val queue = readyQueues[i]
-      queue.cancelAllAndDecide()
-      if (queue.futureTasks.isEmpty()) {
-        readyQueues.removeAt(i)
+    synchronized(this) {
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/concurrent/TaskRunner.kt:246]: Please replace `synchronized`  with this:
```
    this.assertThreadHoldsLock()
```
Our only production caller holds a lock when it calls this; I don’t want re-entrant synchronized blocks if I can avoid it.

Possibly also requires us to change `OkHttpClientTestRule` to acquire the lock before making this call.

In either case, DO NOT call APIs in `okhttp3.internal`; they’re undocumented, subject to change, and not for end users’ use.

**Review by swankjesse — CHANGES_REQUESTED** [review]

File: okhttp/src/main/kotlin/okhttp3/internal/concurrent/TaskRunner.kt (hunk 1 above)
Comment (JarvisGG) [reply okhttp/src/main/kotlin/okhttp3/internal/concurrent/TaskRunner.kt]: Ok.

**Review by swankjesse — APPROVED** [review]

**Comment by swankjesse** [comment]
Thank you!

**Pull request merged**